collectors/monitoring_collector.go

```diff
@@ -333,15 +276,121 @@ func (c *MonitoringCollector) reportMonitoringMetrics(ch chan<- prometheus.Metri
+			filter := fmt.Sprintf("metric.type=\"%s\"", metricDescriptor.Type)
+			if c.monitoringDropDelegatedProjects {
+				filter = fmt.Sprintf(
+					"project=\"%s\" AND metric.type=\"%s\"",
+					c.projectID,
+					metricDescriptor.Type)
+			}
+
+			if c.metricsIngestDelay &&
+				metricDescriptor.Metadata != nil &&
+				metricDescriptor.Metadata.IngestDelay != nil {
+				ingestDelay := metricDescriptor.Metadata.IngestDelay.AsDuration()
+				level.Debug(c.logger).Log("msg", "adding ingest delay", "descriptor", metricDescriptor.Type, "delay", ingestDelay)
+				endTime = endTime.Add(ingestDelay * -1)
+				startTime = startTime.Add(ingestDelay * -1)
+			}
+
+			for _, ef := range c.metricsFilters {
+				if strings.Contains(metricDescriptor.Type, ef.Prefix) {
+					filter = fmt.Sprintf("%s AND (%s)", filter, ef.Modifier)
+				}
+			}
+
+			level.Debug(c.logger).Log("msg", "retrieving Google Stackdriver Monitoring metrics with filter", "filter", filter)
+
+			request := &monitoringpb.ListTimeSeriesRequest{
+				Name:   utils.ProjectResource(c.projectID),
+				Filter: filter,
+				Interval: &monitoringpb.TimeInterval{
+					EndTime: &timestamppb.Timestamp{
+						Seconds: endTime.Unix(),
+						Nanos:   0,
+					},
+					StartTime: &timestamppb.Timestamp{
+						Seconds: startTime.Unix(),
+						Nanos:   0,
+					},
+				},
+				View: monitoringpb.ListTimeSeriesRequest_FULL,
+			}
+			ctx, cancel := context.WithTimeout(context.Background(), c.timeout)
+			defer cancel()
+			it := c.metricClient.ListTimeSeries(ctx, request)
+
+			var results []*monitoringpb.TimeSeries
+			apiCalls := 1.0
+			for {
+				// There's nothing exposed in https://pkg.go.dev/google.golang.org/api/iterator@v0.103.0 which lets you
+				// know an API call was initiated. You might think https://pkg.go.dev/google.golang.org/api/iterator@v0.103.0#NewPager
+				// could do it but the pageSize is a superficial page size that the consumer sets and has impact on API call paging.
+				// If we know there are no items left in the current page and there's a non-empty page token then calling Next() is going to initiate an API call
+				if it.PageInfo().Remaining() == 0 && it.PageInfo().Token != "" {
+					apiCalls += 1.0
+				}
+				timeSeries, err := it.Next()
+				if err == iterator.Done {
+					break
+				}
+				if err != nil {
+					level.Error(c.logger).Log("msg", "error retrieving Time Series metrics for descriptor", "descriptor", metricDescriptor.Type, "err", err)
+					errChannel <- err
+					break
+				}
+				results = append(results, timeSeries)
+			}
+			c.apiCallsTotalMetric.Add(apiCalls)
+
+			if err := c.reportTimeSeriesMetrics(results, metricDescriptor, ch, begun); err != nil {
+				level.Error(c.logger).Log("msg", "error reporting Time Series metrics for descriptor", "descriptor", metricDescriptor.Type, "err", err)
+				errChannel <- err
+			}
```

Comment:
Similar behavior difference with iterating to collect all vs collecting by page. This one is more frustrating because the page limiter isn't the TimeSeries it's the amount of points returned which are embedded in the timeseries. So, even if using the Pager it's going to try to page by TimeSeries so setting the page size in the Pager implementation doesn't translate to the way the API pages.